sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala

```diff
@@ -94,36 +95,61 @@ object ExtractPythonUDFFromAggregate extends Rule[LogicalPlan] {
  */
 object ExtractPythonUDFs extends Rule[SparkPlan] with PredicateHelper {
 
-  private def hasPythonUDF(e: Expression): Boolean = {
+  private case class EvalTypeHolder(private var evalType: Int = -1) {
```

Comment:
How about this:

```scala
  private type EvalType = Int
  private type EvalTypeChecker = Option[EvalType => Boolean]

  private def collectEvaluableUDFsFromExpressions(expressions: Seq[Expression]): Seq[PythonUDF] = {
    // Eval type checker is set in the middle of checking because once it's found,
    // the same eval type should be checked .. blah blah
    var evalChecker: EvalTypeChecker = None

    def collectEvaluableUDFs(expr: Expression): Seq[PythonUDF] = expr match {
      case udf: PythonUDF if PythonUDF.isScalarPythonUDF(udf) && canEvaluateInPython(udf)
        && evalChecker.isEmpty =>
        evalChecker = Some((otherEvalType: EvalType) => otherEvalType == udf.evalType)
        collectEvaluableUDFs(expr)
      case udf: PythonUDF if PythonUDF.isScalarPythonUDF(udf) && canEvaluateInPython(udf)
        && evalChecker.get(udf.evalType) =>
        Seq(udf)
      case e => e.children.flatMap(collectEvaluableUDFs)
    }

    expressions.flatMap(collectEvaluableUDFs)
  }

  def apply(plan: SparkPlan): SparkPlan = plan transformUp {
    case plan: SparkPlan => extract(plan)
  }

  /**
   * Extract all the PythonUDFs from the current operator and evaluate them before the operator.
   */
  private def extract(plan: SparkPlan): SparkPlan = {
    val udfs = collectEvaluableUDFsFromExpressions(plan.expressions)
```

Reply:
I see... You uses a var and nested function definition and var to remove the need of a holder object. 

IMHO I usually find nested function definition and function that refers to variable outside its definition scope hard to read, but it could be my personal preference. 

Another thing I like about the current impl the is `EvalTypeHolder` class ensures its value is ever changed once it's set so I think that's more robust.

That being said, I am ok with your suggestions too if you insist or @BryanCutler also prefers it.

Reply:
yup. I do avoid nested functions but I found here is where is's needed. If it's clear when it's set and unset within a function, I think the shorter one is fine.

Reply:
Ok, I will update the code then.